### ISSUE TITLE
embedded scripts from the 5.0.0-alpha7 source distribution do not work with VS2010

### ISSUE THREAD

**Issue by manxorist** [premake/premake-core#384]
Using https://github.com/premake/premake-core/releases/download/v5.0.0-alpha7/premake-5.0.0-alpha7-src.zip with Visual Studio 2010 (after working around issue #383 either by directly patching the project files to the older toolset version or by running `premake5.exe vs2010` (using the premake5 binary from https://github.com/premake/premake-core/releases/download/v5.0.0-alpha7/premake-5.0.0-alpha7-windows.zip)) results in a non-functioning premake5.exe:
```
C:\Users\manx\projects\openmpt\tmp\premake-5.0.0-alpha7>bin\release\premake5.exe

Error: src/_premake_main.lua: bad header in precompiled chunk
```

The problem goes away once I use the provided premake5 binary to execute `premake5.exe embed` and then compile with Visual Studio 2010.
This additional step was not required with 5.0.0-alpha6 and makes it very difficult to compile premake from the source distribution without resorting to an already compiled premake5.exe.

**Comment by manxorist** [comment]
Using the just compiled premake5.exe (with the non-functioning scripts) to embed the scripts again by calling `bin\release\premake5.exe --scripts=. embed` (and thereby not using the embedded scripts in this call), and then compiling again, also works.

**Comment by tvandijck** [comment]
The "bootstrap.mak" file should really be used to make a premake5.exe binary.. this builds premake5 from source using it own source code, instead of relying on a pre-existing premake5 binary. Some features in the latest embed process are simply not working with pre-existing premake binaries.

using the bootstrap.mak file is pretty simple:
```
nmake -f Bootstrap.mak windows
```

**Comment by manxorist** [comment]
Well, documentation at https://github.com/premake/premake-core/wiki/Building-Premake#using-a-source-code-package says that all that should be required is building using the project files, and that simply does not work.
`Bootstrap.mak` is not even distributed in the `premake-5.0.0-alpha7-src.zip` (which I think is correct and consistent with documentation (i.e. what would be the point in distributing actual project files when there is no way to use them in the first place?)).

**Comment by starkos** [comment]
The documentation is correct, it should be possible to use the pre-generated project files to build Premake with your tool of choice.

I think the root cause of this problem is described in [issue 383](https://github.com/premake/premake-core/issues/383): the contrib projects aren't respecting the `--to` argument and so are getting overwritten by the packaging script.

**Comment by manxorist** [comment]
Well, issue #383 is another problem but definitely completely distinct from this one.

As I already noted in the original report here, this issue is solely about the embedded scripts that come with the source distribution zip file not working with MSVC. The embedded scripts do not work when premake gets compiled with MSVC.
I'm no Lua expert, but to me it looks like that previously the lua scripts were embedded as raw lua text strings and are now embedded as some lua bytecode. Whether this is the root cause of the problem I cannot tell, but the error message also somewhat hints in that direction.
For reference, the embedded scripts as distributed actually do work when compiled with gcc 4.8 on Ubuntu 14.04 (x86-64) (Note: I experienced the same problem as reported in issue #372 when using `build/gmake.unix/Makefile`. That one actually seems related to #383).

**Comment by starkos** [comment]
Okay, just tested, I can reproduce this locally. No idea why it is having an issue right now, but I know that Lua byte code is known to have some portability issues. Short term workaround is to add the `--no-bytecode` flag when embedding, but obviously that's not a good solution.

**Comment by tvandijck** [comment]
When I introduced the bytecode, I did mention that 32 bit and 64 bit bytecode are actually different... so depending on the platform where the embed was executed (on windows it will generate 32 bit bytecode for example), the script.c will need to be generated on the platform where it is used.

So the package.lua script CANNOT make a script.c that works across all platforms, simply because if you run the package.lua script on ubuntu 64-bit, it will make a 64bit bytecode version, that won't work on 32bit windows, and visa versa...

The package.lua script should really make a scripts.c that uses the "--no-bytecode" flag to make sure a source distribution is platform independent... the binary distributions however do not have to use that, since those binaries are either 64 or 32 bit and will have proper bytecode embedded in them. Using 32 bit bytecode on 64 bit will result in the errors given, the other way around as well.

**Comment by starkos** [comment]
Ok nice, so it is that easy. I'll make sure that happens before the next alpha build.

**Comment by tvandijck** [comment]
In all honesty, I think the bootstrap.mak should be the recommended way, and distributing pre-built projects/makefiles as part of the source distribution is confusing, and leads to errors we don't encountered before, since we ourselves never really use it...

I had never heard of the pre-built project files before I responded to this particular issue, which is why I mentioned to use the bootstrap process. What is the reason we distribute pre-built projects? maybe that was a legacy thing that with the bootstrap process is no longer needed?

**Comment by starkos** [comment]
I'm okay with that, in theory. We'll need to get [issue 389](https://github.com/premake/premake-core/issues/389) resolved.

**Comment by manxorist** [comment]
The proposed solution for resolving the issue sounds fine for me. 

Regarding the generated project files: Well, this is getting somewhat off-topic for the original issue, but in my opinion, distributing project files is very useful.
Premake is a tool to generate native project files for various platforms. If it does not itself come with such self-generated project files, it would give a rather bad first impression to first time users and require users to just learn yet another (albeit simple) new build system (namely the premake bootstrap build system). Premake's whole point is avoiding exactly this kind of thing and having things work with the native build systems that anyone is already familiar with anyway.

**Comment by starkos** [comment]
@tvandijck would you be opposed to switching the default to `--no-bytecode`? That way people who aren't familiar with this won't get themselves into trouble. (Including me, apparently, who totally forgot about the 32-/64-bit thing even though you had told me about it.)

We can switch the flag to `--bytecode` instead (which, as I recall, was how you set all of this up originally and I asked you to switch it) and use it on the official builds.

If you are okay with that, I can look into making the change. I'd like to get things patched up and push another alpha to get this all sorted (along with fixing the location of the contrib libraries).

**Comment by tvandijck** [comment]
Yeah, I have no objection.. it's what I originally did indeed... I'll just add the --bytecode to our deploy scripts, but that is easy to do on our end. That said... you do want to make sure the test server runs with both, so both get tested.

**Comment by starkos** [comment]
This is fixed in PR #427.
